src/test/ui/asm/issue-51431.rs

```diff
@@ -1,3 +1,4 @@
+// build-fail
```

Comment:
I don't see how this is checked.
Looks like if the test starts failing at the "check" phase it will still pass with the `// build-fail` annotation.

Why are such annotations useful then?

Reply:
Also, how the `// build-fail` test set was collected?
This can be done by introducing some kind of analogue of `--pass check`, but for failing tests, but I don't see it in this PR.

Reply:
>  Looks like if the test starts failing at the "check" phase it will still pass with the `// build-fail` annotation.

Ouch; Didn't think of that. To fix it, I think we'd need to compile the test twice, once as `check-pass` and once as `build-fail`.

> Also, how the `// build-fail` test set was collected?

A set of tests started failing because they did not fail with `--emit metadata`.

>  This can be done by introducing some kind of analogue of `--pass check`, but for failing tests, but I don't see it in this PR.

Didn't feel necessary, but I can add `--fail` if you think it's useful.

Reply:
>but I can add --fail if you think it's useful.

No, no, just running twice would probably be preferable, the number of such tests is not too large, but we need to measure.
This is material for another PR anyway.

Reply:
I think I fixed it, but I had to refactor things to get rid of some implicit state to make it work. I think things should be less brittle & more understandable now. Let's see what the CI builder says. :)